File: 139.单词拆分.java
/*
 * @lc app=leetcode.cn id=139 lang=java
 *
 * [139] 单词拆分
 */

// @lc code=start

import java.util.*;
class Solution {
    private Map<String,Boolean> memx = new HashMap<>();
    public boolean wordBreak(String s, List<String> wordDict) {
        return backtrack(s, new HashSet<>(wordDict));

    }

    boolean backtrack(String s, Set<String> wordDict){
        if(memx.containsKey(s))return memx.get(s);
        if(s.length()==0)return true;
        boolean res = false;
        for(int i=0;i<s.length();i++){
            if(wordDict.contains(s.substring(0, i+1))){
                res = res || backtrack(s.substring(i+1, s.length()), wordDict);
            }
        }
        memx.put(s,res);
        return res;

    }
}
// @lc code=end

File: 139.单词拆分.py
#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memx = {}
        def backtrack(s):
            if s in memx: return memx[s]
            if len(s) == 0: return True
            res = False
            for i in range(len(s)):
                if s[:i+1] in wordDict:
                    res = backtrack(s[i+1:]) or res
            memx[s] = res
            return res
        return backtrack(s)
            
# @lc code=end

